File: 2026/2225/2225f.py
class Hash:
    def __init__(self, s):
        self.n = len(s)

        self.base1 = 91138233
        self.base2 = 97266353
        self.mod1 = 10**9 + 7
        self.mod2 = 10**9 + 9

        self.h1 = [0] * (self.n + 1)
        self.h2 = [0] * (self.n + 1)
        self.p1 = [1] * (self.n + 1)
        self.p2 = [1] * (self.n + 1)

        for i, ch in enumerate(s):
            x = ord(ch)
            self.h1[i + 1] = (self.h1[i] * self.base1 + x) % self.mod1
            self.h2[i + 1] = (self.h2[i] * self.base2 + x) % self.mod2
            self.p1[i + 1] = (self.p1[i] * self.base1) % self.mod1
            self.p2[i + 1] = (self.p2[i] * self.base2) % self.mod2

    def get(self, l, r):
        x1 = (self.h1[r] - self.h1[l] * self.p1[r - l]) % self.mod1
        x2 = (self.h2[r] - self.h2[l] * self.p2[r - l]) % self.mod2
        return (x1, x2)

def lcp(rh, i, j, max_len):
    lo, hi = 0, max_len
    while lo < hi:
        mid = (lo + hi + 1) // 2
        if rh.get(i, i + mid) == rh.get(j, j + mid):
            lo = mid
        else:
            hi = mid - 1
    return lo


t = int(input())
for _ in range(t):
    n, l, k = map(int, input().split())
    s = input()

    limit = n - k * l
    if limit < 0:
        print("NO")
        continue

    if k == 1:
        print("YES")
        print(s)
        continue

    rh = Hash(s)

    result = 0
    result_len = l + limit

    for i in range(l, n - l + 1, l):
        upper = min(i + limit + 1, n - l + 1)
        for j in range(i, upper):
            cur_len = l + limit - (j % l)
            cmp_len = min(
                result_len,
                cur_len,
                n - result,
                n - j
            )


            c = lcp(rh, result, j, cmp_len)
            if c < cmp_len and s[j + c] > s[result + c]:
                result = j
                result_len = cur_len

    print("YES")
    print(s[result : result + result_len])

File: 2026/2225/2225f.cpp
/*!
 * author: manhhomienbienthuy
 * created: 2026-06-30T04:59:18+00:00
 * https://codeforces.com/contest/2225/problem/F
 */

#include <bits/stdc++.h>
using namespace std;

const int MOD = 1000000007;
const int BASE = 91138233;

int main() {
  int t;
  cin >> t;
  while (t--) {
    int n, l, k;
    cin >> n >> l >> k;
    string s;
    cin >> s;

    if (1ll * k * l > n) {
      cout << "NO\n";
      continue;
    }

    if (k == 1) {
      cout << "YES\n";
      cout << s << "\n";
      continue;
    }

    // rolling hash
    vector<int64_t> h(n + 1, 0), p(n + 1, 1);
    for (int i = 0; i < n; i++) {
      int x = s[i];
      h[i + 1] = (h[i] * BASE + x) % MOD;
      p[i + 1] = (p[i] * BASE) % MOD;
    }

    int limit = n - k * l;
    int result = 0;
    int result_len = l + limit;

    for (int i = l; i < n - l + 1; i++) {
      if (i % l > limit) continue;
      int cur_len = min(n - i, l + limit - (i % l));
      int cmp_len = min(result_len, cur_len);
      int lo = 0, hi = cmp_len;
      while (lo < hi) {
        int mid = (lo + hi + 1) / 2;
        int64_t ga = (h[result + mid] - h[result] * p[mid] % MOD + MOD) % MOD,
                gb = (h[i + mid] - h[i] * p[mid] % MOD + MOD) % MOD;
        if (ga == gb)
          lo = mid;
        else
          hi = mid - 1;
      }
      if ((lo < cmp_len && s[i + lo] > s[result + lo]) || lo == result_len) {
        result = i;
        result_len = cur_len;
      }
    }

    cout << "YES\n";
    cout << s.substr(result, result_len) << "\n";
  }
  return 0;
}
